File: 07.Portals/Storefront/AngJS/framework/store/_core/serverSidePaging.ts
module cef.store.core {
    export class ServerSidePaging<
            TRecord extends api.BaseModel,
            TPagedRecord extends api.PagedResultsBase<TRecord>> {
        // Data
        protected _totalCount = 0;
        protected _filteredCount = null;
        protected _quickFilter: string = "";
        protected _pageSize = 8;
        protected _data: { [page: number]: TRecord[] } = { };
        protected _dataUnpaged: TRecord[] = [];
        protected _currentPage = 0;
        protected _cachedFilteredData: { [page: number]: TRecord[] } = null;
        protected _cachedAllPages: number[] = null;
        pagesLoaded: number[] = [];
        get totalCount(): number {
            return this._totalCount;
        }
        get filteredCount(): number {
            if (!this._filteredCount) {
                this.filteredData; // Fills the value
            }
            return this._filteredCount;
        }
        get pageSize(): number {
            return this._pageSize;
        }
        set pageSize(value: number) {
            this._pageSize = value;
            this.pagesLoaded = [];
            this.resetCaches();
            this.resetAll();
            this.search();
        }
        get quickFilter(): string {
            return this._quickFilter;
        }
        set quickFilter(value: string) {
            this._quickFilter = value;
            this.pagesLoaded = [];
            this.resetCaches();
            this.search();
        }
        get dataUnpaged(): TRecord[] {
            return this._dataUnpaged;
        }
        get data(): { [page: number]: TRecord[] } {
            return this._data;
        }
        set data(value: { [page: number]: TRecord[] }) {
            if (value) {
                let ttlCount = 0;
                const unpaged = [];
                this.pagesLoaded = [];
                Object.keys(value).forEach(page => {
                    ttlCount = ttlCount + value[Number(page)].length;
                    value[Number(page)].forEach(i => {
                        i["__page"] = Number(page);
                        unpaged.push(i);
                    });
                    this.pagesLoaded.push(Number(page));
                });
                this._data = value;
                this._dataUnpaged = unpaged;
                this._totalCount = ttlCount;
            } else {
                this._data = { };
                this._dataUnpaged = [];
                this.pagesLoaded = [];
                this._totalCount = 0;
            }
            this.resetCaches();
        }
        get currentPageFrom(): number {
            return this.skip() + 1;
        }
        get currentPageTo(): number {
            const unlimitedTo = this.skip() + this.pageSize;
            return this.filteredCount < unlimitedTo
                ? this.filteredCount
                : unlimitedTo;
        }
        get showingLabelContent(): string {
            // NOTE: Translations are an issue due to use in both admin and storefront
            return this.filteredCount <= 0
                ? "No Results"
                : this.$filter
                    && `Showing ${this.$filter("number")(this.currentPageFrom, 0)} to ${
                            this.$filter("number")(this.currentPageTo, 0)} of ${
                                this.$filter("number")(this.filteredCount, 0)}`;
        }
        get currentPage(): number {
            if (angular.isUndefined(this._currentPage)) {
                this._currentPage = 0;
            }
            return this._currentPage;
        }
        set currentPage(newValue: number) {
            const newValueChecked = Math.max(0, Math.min(newValue || 0, this.maxPage() - 1)); // -1 to make it zero based
            if (newValueChecked !== 0 && this._currentPage === newValueChecked) {
                return;
            }
            this._currentPage = newValueChecked;
            this.search().then(() => { /* Do Nothing*/ });
        }
        // Cache Management
        resetCachedPages(): void {
            this._cachedAllPages = null;
        }
        resetCachedFilteredData(): void {
            this._cachedFilteredData = null;
            this._filteredCount = 0;
        }
        resetCaches(): void {
            this.resetCachedFilteredData();
            this.resetCachedPages();
            this._currentPage = this.currentPage;
        }
        resetAll(): void {
            this.data = { };
        }
        // Calculate
        setCurrentPageIfDifferent(newValue: number): void {
            if (angular.isUndefined(newValue)) {
                newValue = 0;
            }
            if (this.currentPage === newValue) {
                return;
            }
            this.currentPage = newValue;
        }
        skip(): number {
            return this.currentPage * this.pageSize;
        }
        maxPage(): number {
            return Math.ceil(this.filteredCount / this.pageSize);
        }
        /**
         * @deprecated Use `pages()` instead. Here for compatibility with
         * the WithSets version of this class
         */
        allPages(): number[] {
            if (this._cachedAllPages) {
                return this._cachedAllPages;
            }
            this.pages(); // Populates the data
            return this._cachedAllPages;
        }
        pages(): number[] {
            if (this._cachedAllPages) {
                return this._cachedAllPages;
            }
            const all = new Array<number>();
            for (let i = 0; i < this.maxPage(); i++) {
                all.push(i);
            }
            this._cachedAllPages = all;
            return this._cachedAllPages;
        }
        get filteredData(): { [page: number]: TRecord[] } {
            if (this._cachedFilteredData) {
                return this._cachedFilteredData;
            }
            if (!this.data || !this.$filter) {
                this._filteredCount = 0;
                this._cachedFilteredData = { [0]: [] };
                return this._cachedFilteredData;
            }
            if (!this.quickFilter) {
                this._cachedFilteredData = this._data;
                this._filteredCount = this._totalCount;
                return this._cachedFilteredData;
            }
            const rawFiltered = this.$filter("filter")(this._dataUnpaged, { "$": this.quickFilter });
            if (!rawFiltered.length) {
                this._filteredCount = 0;
                this._cachedFilteredData = { [0]: [] };
                return this._cachedFilteredData;
            }
            const paged: { [page: number]: TRecord[] } = { };
            let pageIndex = 0;
            rawFiltered.forEach(i => {
                if (!paged[pageIndex]) {
                    paged[pageIndex] = [];
                }
                if (paged[pageIndex].length >= this._pageSize) {
                    pageIndex++;
                }
                if (!paged[pageIndex]) {
                    paged[pageIndex] = [];
                }
                i["__filteredPage"] = pageIndex;
                paged[pageIndex].push(i);
            });
            this._cachedFilteredData = paged;
            this._filteredCount = this._totalCount = rawFiltered.length;
            return this._cachedFilteredData;
        }
        // Navigate
        goToFirstPage = () => {
            this.setCurrentPageIfDifferent(0);
        }
        goToPreviousPage = () => {
            this.setCurrentPageIfDifferent(Math.max(0, this.currentPage - 1));
        }
        goToNextPage = () => {
            this.setCurrentPageIfDifferent(Math.min(this.maxPage(), this.currentPage + 1));
        }
        goToLastPage = () => {
            this.setCurrentPageIfDifferent(this.maxPage() - 1); // -1 to make it zero based
        }
        // Functions
        search(): ng.IPromise<TRecord[]> {
            if (angular.isUndefined(this._currentPage)) {
                return this.$q.reject("No current page");
            }
            if (this.pagesLoaded.indexOf(this._currentPage) !== -1) {
                return this.$q.resolve(this._data[this._currentPage]);
            }
            return this.$q((resolve, reject) => {
                if (angular.isFunction(this.waitCondition)) {
                    const wait = this.waitCondition();
                    if (wait) {
                        reject("Waiting");
                        return;
                    }
                }
                let dto = <api.BaseSearchModel>{
                    Paging: <api.Paging>{
                        StartIndex: this._currentPage + 1, // Skip
                        Size: (this.pageSize || 8) // Take
                    }
                }
                if (angular.isFunction(this.searchParamsToMerge)) {
                    const toMerge = this.searchParamsToMerge();
                    if (toMerge && Object.keys(toMerge).length > 0) {
                        dto = angular.merge(dto, toMerge);
                    }
                }
                if (this.quickFilter) {
                    dto[this.searchParameterName || "IDOrCustomKeyOrName"] = this.quickFilter;
                }
                this.searchCall(dto).then(r => {
                    if (!this._data) {
                        this._data = [];
                    }
                    if (!this.data[this._currentPage]) {
                        this._data[this._currentPage] = [];
                    }
                    r.data.Results.forEach(i => {
                        i["__page"] = this._currentPage;
                        this._data[this._currentPage].push(i);
                        this.dataUnpaged.push(i);
                    });
                    this._data[this._currentPage] = _.uniqBy(this._data[this._currentPage], x => x.ID).sort(x => x.ID);
                    this._dataUnpaged = _.uniqBy(this._dataUnpaged, x => x.ID).sort(x => x.ID);
                    this.pagesLoaded.push(this._currentPage);
                    this._totalCount = r.data.TotalCount;
                    this.resetCaches();
                    this.filteredData;
                    resolve(this._data);
                    if (this.$rootScope) {
                        this.$rootScope.$broadcast(this.cvServiceStrings.events.paging.searchComplete,
                            this.name,
                            this._data[this._currentPage]);
                    }
                }).catch(reject);
            });
        }
        // Constructor
        constructor(
                private readonly $rootScope: ng.IRootScopeService,
                readonly $scope: ng.IScope,
                private readonly $filter: ng.IFilterService,
                private readonly $q: ng.IQService,
                private readonly cvServiceStrings: services.IServiceStrings,
                private readonly searchCall:
                    (searchModel: api.BaseSearchModel) => ng.IHttpPromise<TPagedRecord>,
                size: number,
                private readonly name: string,
                private readonly searchParameterName: string = "IDOrCustomKeyOrName",
                private readonly searchParamsToMerge: () => object = () => { return { Active: true, AsListing: true }; },
                private readonly waitCondition: () => boolean = null) {
            this._pageSize = size;
            this._currentPage = 0;
            this.search();
            const unbind1 = $scope.$watch(() => angular.toJson(searchParamsToMerge()), () => this.search());
            $scope.$on(this.cvServiceStrings.events.$scope.$destroy, () => {
                if (angular.isFunction(unbind1)) { unbind1(); }
            });
        }
    }
}

File: 07.Portals/XPortal/Client/BrandAdminPortal/Shared/Paged/ServerSidePaging.cs
// <copyright file="ServerSidePaging.cs" company="clarity-ventures.com">
// Copyright (c) 2022-2023 clarity-ventures.com. All rights reserved.
// </copyright>
// <summary>Implements the server side paging class</summary>
namespace Clarity.Ecommerce.UI.XPortal.SharedLibrary.Shared.Paged
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Threading.Tasks;
    using MVC.Api.Models;
    using MVC.Core;
    using MVC.Utilities;

    /// <summary>A server side paging.</summary>
    /// <typeparam name="TRecord">     Type of the record.</typeparam>
    /// <typeparam name="TPagedRecord">Type of the paged record collection wrapper.</typeparam>
    /// <typeparam name="TEndpoint">   Type of the endpoint.</typeparam>
    public class ServerSidePaging<TRecord, TPagedRecord, TEndpoint>
        where TRecord : BaseModel
        where TPagedRecord : PagedResultsBase<TRecord>
        where TEndpoint : BaseSearchModel, new()
    {
        // Data
        private int totalCount;
        private int? filteredCount;
        private string? quickFilter = string.Empty;
        private int pageSize;
        private Dictionary<int/*page*/, List<TRecord>> data = new();
        private List<TRecord> dataUnpaged = new();
        private int currentPage;
        private Dictionary<int/*page*/, List<TRecord>>? cachedFilteredData;
        private int[]? cachedAllPages;

        private readonly Func<TEndpoint, Task<IHttpPromiseCallbackArg<TPagedRecord>>> searchCall;
        private readonly string? searchParameterName;
        private readonly Func<TEndpoint?>? searchParamsToMerge;
        private readonly Func<bool>? waitCondition;
        private readonly Action? callback;

        /// <summary>The pages loaded.</summary>
        public List<int> PagesLoaded = new();

        /// <summary>(Immutable) The name.</summary>
        public readonly string Name;

        /// <summary>Initializes a new instance of the <see cref="ServerSidePaging{TRecord, TPagedRecord, TEndpoint}"/>
        /// class.</summary>
        /// <param name="searchCall">         The search call.</param>
        /// <param name="size">               The size.</param>
        /// <param name="name">               The name.</param>
        /// <param name="searchParameterName">Name of the search parameter.</param>
        /// <param name="searchParamsToMerge">The search parameters to merge.</param>
        /// <param name="waitCondition">      The wait condition.</param>
        public ServerSidePaging(
            Func<TEndpoint, Task<IHttpPromiseCallbackArg<TPagedRecord>>> searchCall,
            int size,
            string name,
            string? searchParameterName = "IDOrCustomKeyOrName",
            Func<TEndpoint>? searchParamsToMerge = null,
            Func<bool>? waitCondition = null,
            Action? callback = null)
        {
            searchParamsToMerge ??= () => new TEndpoint { Active = true, AsListing = true };
            pageSize = size;
            currentPage = 0;
            this.searchParamsToMerge = searchParamsToMerge;
            Name = name;
            this.searchParameterName = searchParameterName;
            this.waitCondition = waitCondition;
            this.searchCall = searchCall;
            this.callback = callback;
            Search();
            /* TODO: Convert this to C#
            const unbind1 = $scope.$watch(() => angular.toJson(searchParamsToMerge()), () => this.search());
            $scope.$on(this.cvServiceStrings.events.$scope.$destroy, () => {
                if (angular.isFunction(unbind1)) { unbind1(); }
            });
            */
        }

        /// <summary>Gets the number of totals.</summary>
        /// <value>The total number of count.</value>
        public int TotalCount => totalCount;

        /// <summary>Gets the number of filtered.</summary>
        /// <value>The number of filtered.</value>
        public int FilteredCount
        {
            get
            {
                if (filteredCount is null)
                {
                    _ = FilteredData; // Fills the value
                }
                return filteredCount!.Value;
            }
        }

        /// <summary>Gets or sets the size of the page.</summary>
        /// <value>The size of the page.</value>
        public int PageSize
        {
            get => pageSize;
            set
            {
                pageSize = value;
                PagesLoaded.Clear();
                ResetCaches();
                ResetAll();
                Search();
            }
        }

        /// <summary>Gets or sets the quick filter.</summary>
        /// <value>The quick filter.</value>
        public string? QuickFilter
        {
            get => quickFilter;
            set
            {
                quickFilter = value;
                PagesLoaded.Clear();
                ResetCaches();
                Search();
            }
        }

        /// <summary>Gets the data unpaged.</summary>
        /// <value>The data unpaged.</value>
        public List<TRecord> DataUnpaged => dataUnpaged;

        /// <summary>Gets or sets the data.</summary>
        /// <value>The data.</value>
        public Dictionary<int, List<TRecord>> Data
        {
            get => data;
            set
            {
                if (value != null!)
                {
                    var ttlCount = 0;
                    List<TRecord> unpaged = new();
                    PagesLoaded.Clear();
                    foreach (var page in value.Keys)
                    {
                        ttlCount += value[page].Count;
                        foreach (var i in value[page])
                        {
                            // TODO: i["__page"] = page;
                            unpaged.Add(i);
                        }
                        PagesLoaded.Add(page);
                    }
                    data = value;
                    dataUnpaged = unpaged;
                    totalCount = ttlCount;
                }
                else
                {
                    data = new();
                    dataUnpaged.Clear();
                    PagesLoaded.Clear();
                    totalCount = 0;
                }
                ResetCaches();
            }
        }

        /// <summary>Gets the current page from.</summary>
        /// <value>The current page from.</value>
        public int CurrentPageFrom => Skip() + 1;

        /// <summary>Gets the current page to.</summary>
        /// <value>The current page to.</value>
        public int CurrentPageTo
        {
            get
            {
                var unlimitedTo = Skip() + PageSize;
                return FilteredCount < unlimitedTo
                    ? FilteredCount
                    : unlimitedTo;
            }
        }

        /// <summary>Gets the showing label content.</summary>
        /// <value>The showing label content.</value>
        public string ShowingLabelContent => FilteredCount <= 0
            ? "No Results"
            : $@"Showing {CurrentPageFrom:N0} to {CurrentPageTo:N0} of {FilteredCount:N0}";

        /// <summary>Gets or sets the current page.</summary>
        /// <value>The current page.</value>
        public int CurrentPage
        {
            get
            {
                if (currentPage <= 0)
                {
                    currentPage = 0;
                }
                return currentPage;
            }
            set
            {
                var newValueChecked = Math.Max(0, Math.Min(value, MaxPage() - 1)); // -1 to make it zero based
                if (newValueChecked != 0 && currentPage == newValueChecked)
                {
                    return;
                }
                currentPage = newValueChecked;
                Search();
            }
        }

        /// <summary>Resets the cached pages.</summary>
        public void ResetCachedPages()
        {
            cachedAllPages = null;
        }

        /// <summary>Resets the cached filtered data.</summary>
        public void ResetCachedFilteredData()
        {
            cachedFilteredData = null;
            filteredCount = 0;
        }

        /// <summary>Resets the caches.</summary>
        public void ResetCaches()
        {
            ResetCachedFilteredData();
            ResetCachedPages();
            currentPage = CurrentPage;
        }

        /// <summary>Resets all.</summary>
        public void ResetAll()
        {
            Data = new();
        }

        /// <summary>Sets current page if different.</summary>
        /// <param name="newValue">The new value.</param>
        public void SetCurrentPageIfDifferent(int newValue)
        {
            if (newValue < 0)
            {
                newValue = 0;
            }
            if (CurrentPage == newValue)
            {
                return;
            }
            CurrentPage = newValue;
        }

        /// <summary>Gets the skip.</summary>
        /// <returns>An int.</returns>
        public int Skip()
        {
            return CurrentPage * PageSize;
        }

        /// <summary>Maximum page.</summary>
        /// <returns>An int.</returns>
        public int MaxPage()
        {
            return (int)Math.Ceiling(FilteredCount / (decimal)PageSize);
        }

        /**
        * @deprecated Use `pages()` instead. Here for compatibility with
        * the WithSets version of this class
        */
        public int[] AllPages()
        {
            if (cachedAllPages is not null)
            {
                return cachedAllPages;
            }
            Pages(); // Populates the data
            return cachedAllPages!;
        }

        /// <summary>Gets the pages.</summary>
        /// <returns>An int[].</returns>
        public int[] Pages()
        {
            if (cachedAllPages is not null)
            {
                return cachedAllPages;
            }
            var all = new List<int>();
            for (var i = 0; i < MaxPage(); i++)
            {
                all.Add(i);
            }
            cachedAllPages = all.ToArray();
            return cachedAllPages;
        }

        /// <summary>Gets the filtered data set.</summary>
        /// <value>The filtered data set.</value>
        public Dictionary<int, List<TRecord>> FilteredData
        {
            get
            {
                if (cachedFilteredData is not null)
                {
                    return cachedFilteredData;
                }
                if (Contract.CheckEmpty(Data))
                {
                    filteredCount = 0;
                    cachedFilteredData = new()
                    {
                        [0] = new(),
                    };
                    return cachedFilteredData;
                }
                if (!Contract.CheckValidKey(QuickFilter))
                {
                    cachedFilteredData = data;
                    filteredCount = totalCount;
                    return cachedFilteredData;
                }
                // TODO: Use json serialize instead of ToString
                var rawFiltered = dataUnpaged.Where(x => x.ToString()!.Contains(QuickFilter!)).ToList();
                if (Contract.CheckEmpty(rawFiltered))
                {
                    filteredCount = 0;
                    cachedFilteredData = new()
                    {
                        [0] = new(),
                    };
                    return cachedFilteredData;
                }
                Dictionary<int, List<TRecord>> paged = new();
                var pageIndex = 0;
                foreach (var i in rawFiltered)
                {
                    if (!paged.ContainsKey(pageIndex))
                    {
                        paged[pageIndex] = new();
                    }
                    if (paged[pageIndex].Count >= pageSize)
                    {
                        pageIndex++;
                    }
                    if (!paged.ContainsKey(pageIndex))
                    {
                        paged[pageIndex] = new();
                    }
                    // TODO: i["__filteredPage"] = pageIndex;
                    paged[pageIndex].Add(i);
                }
                cachedFilteredData = paged;
                filteredCount = totalCount = rawFiltered.Count;
                return cachedFilteredData;
            }
        }

        /// <summary>Go to first page.</summary>
        public void GoToFirstPage()
        {
            SetCurrentPageIfDifferent(0);
        }

        /// <summary>Go to previous page.</summary>
        public void GoToPreviousPage()
        {
            SetCurrentPageIfDifferent(Math.Max(0, CurrentPage - 1));
        }

        /// <summary>Go to next page.</summary>
        public void GoToNextPage()
        {
            SetCurrentPageIfDifferent(Math.Min(MaxPage(), CurrentPage + 1));
        }

        /// <summary>Go to last page.</summary>
        public void GoToLastPage()
        {
            SetCurrentPageIfDifferent(MaxPage() - 1); // -1 to make it zero based
        }

        /// <summary>Calls for data from the server with any search parameters if set.</summary>
        /// <returns>The results.</returns>
        public void Search()
        {
            if (waitCondition?.Invoke() == true || PagesLoaded.Contains(currentPage))
            {
                return;
            }
            var toMerge = searchParamsToMerge?.Invoke();
            var dto = toMerge ?? new();
            dto.Paging = new()
            {
                StartIndex = currentPage + 1, // Skip
                Size = PageSize, // Take
            };
            if (Contract.CheckValidKey(QuickFilter))
            {
                // TODO: dto[searchParameterName ?? "IDOrCustomKeyOrName"] = QuickFilter;
            }
            var awaiter = searchCall(dto).GetAwaiter();
            awaiter.OnCompleted(() =>
            {
                Console.WriteLine("=============================================");
                var r = awaiter.GetResult();
                if (data == null!)
                {
                    data = new();
                }
                if (!Data.ContainsKey(currentPage))
                {
                    data[currentPage] = new();
                }
                foreach (var i in r.data!.Results!)
                {
                    // TODO: i["__page"] = currentPage;
                    data[currentPage].Add(i);
                    DataUnpaged.Add(i);
                }
                data[currentPage] = data[currentPage].DistinctBy(x => x.ID).OrderBy(x => x.ID).ToList();
                dataUnpaged = dataUnpaged.DistinctBy(x => x.ID).OrderBy(x => x.ID).ToList();
                PagesLoaded.Add(currentPage);
                totalCount = r.data.TotalCount;
                ResetCaches();
                _ = FilteredData;
                callback?.Invoke();
            });
        }
    }
}
